### ISSUE TITLE
Weird AS LOCAL behavior with directories

### ISSUE THREAD

**Issue by jazzdan** [earthly/earthly#703]
Working from this [simple sample repo](https://github.com/jazzdan/earthly-as-local-weirdness), with an Earthfile that looks like this:

```
everythings-flat:
    COPY qux .
    COPY bar .
    SAVE ARTIFACT ./* AS LOCAL out/standalone/

wat:
    COPY qux .
    COPY bar .
    SAVE ARTIFACT . AS LOCAL out/standalone/
```

When I run `earthly +everythings-flat` I would expect to get a multi level file tree in `out/standalone` that mirrors what is in the repo:

```
❯ tree .
.
├── Earthfile
├── bar
│   ├── baz
│   └── goo
│       └── gone
└── qux
    └── foo
```
. Instead what I get is this:

```
❯ earthly +everythings-flat && tree out/standalone
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
No ssh auth socket detected or zero keys loaded; falling back to https for auto auth values
             context | --> local context .
             context | [██████████] transferring .: ... 100%
   +everythings-flat | *cached* --> COPY qux .
   +everythings-flat | *cached* --> COPY bar .
   +everythings-flat | *cached* --> SAVE ARTIFACT ./* +everythings-flat/* AS LOCAL out/standalone/
              output | --> exporting outputs
              output | [██████████] copying files ... 100%
              output | [██████████] sending tarballs ... 100%
=========================== SUCCESS ===========================
   +everythings-flat | Artifact github.com/jazzdan/earthly-as-local-weirdness:main+everythings-flat/baz as local out/standalone/baz
   +everythings-flat | Artifact github.com/jazzdan/earthly-as-local-weirdness:main+everythings-flat/foo as local out/standalone/foo
   +everythings-flat | Artifact github.com/jazzdan/earthly-as-local-weirdness:main+everythings-flat/gone as local out/standalone/gone
out/standalone
├── baz
├── foo
└── gone
```

Only the leaf node files get copied over, none of the directory structure.

If I change the `SAVE ARTIFACT ./*` to a `SAVE ARTIFACT .`, as in `wat` I get:

```
❯ earthly +wat && tree out/standalone
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
No ssh auth socket detected or zero keys loaded; falling back to https for auto auth values
             context | --> local context .
             context | [██████████] transferring .: ... 100%
                +wat | *cached* --> COPY qux .
                +wat | *cached* --> COPY bar .
                +wat | *cached* --> SAVE ARTIFACT . +wat/ AS LOCAL out/standalone/
              output | --> exporting outputs
              output | [██████████] copying files ... 100%
              output | [██████████] sending tarballs ... 100%
=========================== SUCCESS ===========================
                +wat | Artifact github.com/jazzdan/earthly-as-local-weirdness:main+wat/ as local out/standalone
out/standalone
├── baz
├── foo
├── gone
└── index-0
    ├── baz
    ├── foo
    └── goo
        └── gone
```

Which is just weird! What is `index-0`? I guess it's the working directory of Earthly? What happened to the bar directory?

It's possible that we're thinking about how to do this all wrong, and misunderstanding the role of `*` in `SAVE ARTIFACT`. Maybe what we actually want is a `--recursive` flag for `SAVE ARTIFACT`?

cc @hbenl

**Comment by hbenl** [comment]
Maybe the `SAVE AS LOCAL` command should have a `--dir` option like the `COPY` command, just to make it more consistent. If you add this option, please also add a warning to the command's documentation that if you write a `SAVE AS LOCAL` command that references more than one item (either explicitly or using wildcards), at least one of which is a directory, then you most probably want to add the `--dir` option or the contents of several directories get mixed into one.

**Comment by vladaionescu** [comment]
The intention was to always behave as if it had the `--dir` option enabled. This is what the fix we've developed achieves. Hopefully that clears up the behavior. We'll likely make some additional mentions in the docs about the actual behavior.

**Comment by dchw** [comment]
Another thing to note here that seems to have caused some confusion is the `COPY qux .` In a Dockerfile COPY; it doesn't copy `qux`, but instead copies the contents of `qux` into the current directory.

Though weird; I think it was amplified by the bug. Like @vladaionescu said, we are going to add some docs around these semantics so its clearer what to expect.

**Comment by vladaionescu** [comment]
This is now released in `v0.4.5` and reference docs now include some examples about the behavior. Any feedback is appreciated!

**Comment by dchw** [comment]
Here are those docs @vladaionescu mentioned, for posterity:

* https://docs.earthly.dev/earthfile#examples
